### PR TITLE
Cleaning up more algorithms

### PR DESCRIPTION
**Pull request by manu4rhyme-zz**
Added isGoalState() helper methods (and some more minor changes) , to align with the pseudo-code better.

### CODE REVIEW

File: core/src/main/java/aima/core/search/basic/informed/AStarSearch.java [hunk 1]
```diff
@@ -121,6 +121,10 @@ public AStarSearch(ToDoubleFunction<Node<A, S>> h) {
 		return searchController.solution(node);
```
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/informed/AStarSearch.java:121]: Something similar could be done with searchController.solution(node) being replaced by solution(node). i.e.:

	public List<A> solution(Node<A, S> node) {
 		return searchController.solution(node);
 	}

File: core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java [hunk 2]
```diff
@@ -142,22 +142,12 @@ public SuccessorNode getLowestFValueNode(List<SuccessorNode> nodes) {
...
-				lowest = node.f;
-			}
-			else if(node.f < secondLowest) {
-				secondLowest = node.f;
-			}
-		}
-		return secondLowest;
+	public double getSecondLowestFValue(List<SuccessorNode> nodes, Node<A, S> lowest) {		
+		return nodes.stream().filter(x -> x != lowest).min(Comparator.comparingDouble(s -> s.f)).get().f;		
+	}
```
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java]: What if nodes has size() == 1?

File: core/src/main/java/aima/core/search/basic/uninformed/GraphSearch.java [hunk 3]
```diff
@@ -1,110 +1,114 @@
...
-		return searchController.solution(node);
-	}
-
-	public boolean containsState(Queue<Node<A, S>> frontier, Node<A, S> child) {
-		// NOTE: Not very efficient (i.e. linear in the size of the frontier)
-		return frontier.stream().anyMatch(frontierNode -> frontierNode.state().equals(child.state()));
-	}
-}
+package aima.core.search.basic.uninformed;
+
```
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/uninformed/GraphSearch.java]: It looks like you re-formatted the whole file. This makes it extremely difficult to review the changes. Please don't do this.

**Review by ctjoreilly — CHANGES_REQUESTED** [review]
Please address the comments and then squash/rebase as needed.

Thanks

Ciaran

File: core/src/main/java/aima/core/search/basic/informed/AStarSearch.java (hunk 1 above)
Comment (manu4rhyme-zz) [reply core/src/main/java/aima/core/search/basic/informed/AStarSearch.java:121]: Yes, I have already taken this into account. Also for searchController.failure(), nodeFactory.newRootNode() and nodeFactory.newChildNode()

File: core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java (hunk 2 above)
Comment (manu4rhyme-zz) [reply core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java]: True, I overlooked this trivial test case.
Fixed.

File: core/src/main/java/aima/core/search/basic/uninformed/GraphSearch.java (hunk 3 above)
Comment (manu4rhyme-zz) [reply core/src/main/java/aima/core/search/basic/uninformed/GraphSearch.java]: Sorry, it was a white-space issue.
Fixed.

**Comment by manu4rhyme-zz** [comment]
@ctjoreilly Fixed

File: core/src/main/java/aima/core/search/basic/informed/AStarSearch.java (hunk 1 above)
Comment (ctjoreilly) [reply core/src/main/java/aima/core/search/basic/informed/AStarSearch.java:121]: but what about here at line 121 in AStartSearch?

File: core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java [hunk 4]
```diff
@@ -142,22 +142,12 @@ public SuccessorNode getLowestFValueNode(List<SuccessorNode> nodes) {
...
-				lowest = node.f;
-			}
-			else if(node.f < secondLowest) {
-				secondLowest = node.f;
-			}
-		}
-		return secondLowest;
+	public double getSecondLowestFValue(List<SuccessorNode> nodes, SuccessorNode lowest) {		
+		return nodes.stream().filter(x -> x != lowest).min(Comparator.comparing(s -> s.f)).orElse(lowest).f;
+	}
```
Comment (ctjoreilly) [inline core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java:147]: Nice and clean :)

File: core/src/main/java/aima/core/search/basic/informed/AStarSearch.java (hunk 1 above)
Comment (manu4rhyme-zz) [reply core/src/main/java/aima/core/search/basic/informed/AStarSearch.java:121]: But , its inside the solution(node) helper function itself. What can this be replaced by ?

Comment (ctjoreilly) [reply core/src/main/java/aima/core/search/basic/informed/AStarSearch.java:121]: My bad! :) the code collapsing through me off

**Pull request merged**